### ISSUE TITLE
Investigate the memory leak by DocumentHandle within the RenderTree

### ISSUE THREAD

**Issue by Kiyoshika** [gosub-browser/gosub-engine#280]
Somehow the `DocumentHandle` within the `RenderTree` is leaking when running `gosub-bindings/tests/render_tree_test` under valgrind.

May need to use a different type of pointer or wrap the `Document` directly in a `Box`.

**Comment by Sharktheone** [comment]
You said something about the Rc not being leak-safe. I think it's only not leak safe when you have a reference cycle. So you could use `Rc::downgrade` to convert it to `Weak` which is a weak reference and doesn't count as a real reference (so not preventing a de allocation). But i'm not sure if the problem is a reference cycle.

**Comment by Kiyoshika** [comment]
> You said something about the Rc not being leak-safe. I think it's only not leak safe when you have a reference cycle. So you could use `Rc::downgrade` to convert it to `Weak` which is a weak reference and doesn't count as a real reference (so not preventing a de allocation). But i'm not sure if the problem is a reference cycle.

Yes, that’s also what I’ve read but going through the code I didn’t really understand where it was happening

**Comment by Kiyoshika** [comment]
At the moment this is just conjecture, but it's likely the source of the leak is cloning the `DocumentHandle` in each `Node` which is probably not being cleaned up properly and causing lots of weird cyclical references.
